modules/internal/ChapelBase.chpl

```diff
@@ -2375,16 +2375,23 @@ module ChapelBase {
   }
 
 
-  // This is a helper function that I injected to reduce the
-  // compiler's reliance on 'iterable.size' for coforall loops because
-  // we started generating warnings for the return type of
-  // '[range|domain|array].size' changing.
-  //
-  proc chpl_coforallSize(iterable) {
-    if (isRange(iterable) || isDomain(iterable) || isArray(iterable)) then
+  // Support for bounded coforall task counting optimizations
+  proc chpl_supportsBoundedCoforall(iterable, param zippered) param {
+    if zippered && isTuple(iterable) then
+      return chpl_supportsBoundedCoforall(iterable[0], zippered=false);
+    if isRange(iterable) || isDomain(iterable) || isArray(iterable) then
+      return true;
+    return false;
+  }
+
+  proc chpl_boundedCoforallSize(iterable, param zippered) {
+    if !chpl_supportsBoundedCoforall(iterable, zippered) then
+      compilerError("Called chpl_boundedCoforallSize on unsupported type");
+
+    if zippered && isTuple(iterable) then
+      return chpl_boundedCoforallSize(iterable[0], zippered=false);
+    if isRange(iterable) || isDomain(iterable) || isArray(iterable) then
```

Comment:
Could you add an `else` for this?

In the previous code it was just returning `iterable.size`. Looking at the compiler code, I *think* hitting that `else` would be impossible in a healthy compiler, but it'd be nice if the error message is not generic.

Reply:
I dropped the `iterable.size` because we should never hit that code and if we ever do I didn't want to hide that we're calling it with an unexpected type.

Reply:
Oh, but maybe you're just saying throw a halt in there to give a clear error message?

Reply:
Added